certbot-dns-rfc2136/certbot_dns_rfc2136/dns_rfc2136.py

```diff
@@ -0,0 +1,215 @@
+               'RFC 2136 Dynamic Updates.'
+
+    def _validate_algorithm(self, credentials):
+        algorithm = credentials.conf('algorithm')
+        if algorithm:
+            if not self.ALGORITHMS.get(algorithm):
+                raise errors.PluginError("Unknown algorithm: {0}.".format(algorithm))
+
+    def _setup_credentials(self):
+        self.credentials = self._configure_credentials(
+            'credentials',
+            'RFC 2136 credentials INI file',
+            {
+                'name': 'TSIG key name',
+                'secret': 'TSIG key secret',
+                'server': 'The target DNS server'
+            },
+            self._validate_algorithm
+        )
+
+    def _perform(self, domain, validation_name, validation):
+        self._get_rfc2136_client().add_txt_record(domain, validation_name, validation, self.ttl)
+
+    def _cleanup(self, domain, validation_name, validation):
+        self._get_rfc2136_client().del_txt_record(domain, validation_name, validation)
+
+    def _get_rfc2136_client(self):
+        return _RFC2136Client(self.credentials.conf('server'),
+                              self.credentials.conf('name'),
+                              self.credentials.conf('secret'),
+                              self.ALGORITHMS.get(self.credentials.conf('algorithm'),
+                                                  dns.tsig.HMAC_MD5))
+
+
+class _RFC2136Client(object):
+    """
+    Encapsulates all communication with the target DNS server.
+    """
+    def __init__(self, server, key_name, key_secret, key_algorithm):
+        self.server = server
+        self.keyring = dns.tsigkeyring.from_text({
+            key_name: key_secret
+        })
+        self.algorithm = key_algorithm
+
+    def add_txt_record(self, domain_name, record_name, record_content, record_ttl):
+        """
+        Add a TXT record using the supplied information.
+
+        :param str domain: The domain to use to find the closest SOA.
+        :param str record_name: The record name (typically beginning with '_acme-challenge.').
+        :param str record_content: The record content (typically the challenge validation).
+        :param int record_ttl: The record TTL (number of seconds that the record may be cached).
+        :raises certbot.errors.PluginError: if an error occurs communicating with the DNS server
+        """
+
+        domain = self._find_domain(domain_name)
+
+        n = dns.name.from_text(record_name)
+        o = dns.name.from_text(domain)
+        rel = n.relativize(o)
+
+        update = dns.update.Update(
+            domain,
+            keyring=self.keyring,
+            keyalgorithm=self.algorithm)
+        update.add(rel, record_ttl, dns.rdatatype.TXT, record_content)
+
+        try:
+            response = dns.query.tcp(update, self.server)
+            rcode = response.rcode()
+
+            if rcode == dns.rcode.NOERROR:
+                logger.debug('Successfully added TXT record')
```

Comment:
Should other `rcode` values be logged and treated as an error, resulting in a `errors.PluginError` (like in the `except` clause immediately following this)?